src/Executor/ReferenceExecutor.php

```diff
@@ -1094,15 +1095,17 @@ protected function completeAbstractValue(
     protected function defaultTypeResolver($value, $contextValue, ResolveInfo $info, AbstractType $abstractType)
     {
         // First, look for `__typename`.
-        if (
-            $value !== null
-            && (is_array($value) || $value instanceof ArrayAccess)
+        if ((is_array($value) || $value instanceof ArrayAccess)
             && isset($value['__typename'])
             && is_string($value['__typename'])
         ) {
             return $value['__typename'];
         }
```

Comment:
Instead of introducing a new interface, I think we should rather extend this part to have parity with the way the default field resolver accesses keys on the value - see https://github.com/webonyx/graphql-php/blob/03c630d21858eae50898996a40650d6abca3da23/src/Executor/Executor.php#L193-L201

This will allow objects with a property `__typename` - public or virtual - to work, which I think is conceptually much simpler then a separate interface.

Reply:
You are right, it would be more in sync how the rest of the library works.
My rational was that I think an interface makes the intent much clearer.

Im gonna adjust my PR to use the public property approach you proposed.